plugins/destination/postgresql/client/insert.go

```diff
@@ -46,12 +50,19 @@ func (c *Client) InsertBatch(ctx context.Context, messages message.WriteInserts)
 		if _, ok = c.pgTablesToPKConstraints[tableName]; !ok {
 			return fmt.Errorf("table %s not found", tableName)
 		}
-		table := tables.Get(tableName) // will always be present, panic should be produced if not
-		if len(table.PrimaryKeysIndexes()) > 0 {
-			sql = c.upsert(table)
-		} else {
-			sql = c.insert(table)
+
+		sql, ok := queries[tableName]
+		if !ok {
+			// cache the query
+			table := tables.Get(tableName) // will always be present, panic should be produced if not
+			if len(table.PrimaryKeysIndexes()) > 0 {
+				sql = c.upsert(table)
+			} else {
+				sql = c.insert(table)
+			}
+			queries[tableName] = sql
```

Comment:
Do we have any data or even idea of how much time this will save? Based purely on my intuition I am skeptical that it will save any meaningful time in comparison to the latency involved communicating with the DB over the network.

If in fact this does represent a meaningful performance benefit, why not store the queries in the client that way they can be used across multiple batches?

Reply:
Consider large batch sizes (100K+).
If there is only a handful of different tables in the batch, we can win a lot of time by simply reusing query instead of constructing it over & over.

Reply:
I ran a benchmark to see how much performance increase we could expect for a batch with 100k records. Very surprisingly I found that the caching/map version was significantly longer... 
```
=== RUN   TestGenInsert
    insert_test.go:42: With caching enabled took 29.649439459s
    insert_test.go:52: With caching disabled took 12.558788125s
--- PASS: TestGenInsert (183.31s)
```

You can see the code [here](https://github.com/cloudquery/cloudquery/blob/82b4946937b52ad74960502977d108e117ad9fb1/plugins/destination/postgresql/client/insert_test.go)... My intuition was just like yours that the caching would reduce some of the latency, and had no expectation that it would add latency. I would be super interested to learn why it is in fact slower or what I did wrong in the test code!

Reply:
On my machine:
10K
```
=== RUN   TestGenInsert
    insert_test.go:42: With caching enabled took 3.340375ms
    insert_test.go:52: With caching disabled took 103.888792ms
--- PASS: TestGenInsert (6.39s)
PASS
```

One of the issues in the setup you had is that you also measure the alloc time for all of the 100K entries for sql queries.
I've updated the code to use the benchmarks instead in 70c7e80f32091883ef48c7b88fd9315517dca582, here are the results:

#### No caching
Performed by the following command in `plugins/destination/postgresql/client` dir:
```sh
go test \
-test.run=BenchmarkNoCache \
-test.bench=BenchmarkNoCache \
-test.count 10 -test.benchmem -test.benchtime 10000x
```

<details><summary>`BenchmarkNoCache` results</summary>

```
goos: darwin
goarch: arm64
pkg: github.com/cloudquery/cloudquery/plugins/destination/postgresql/client
BenchmarkNoCache-10        10000              9932 ns/op            6016 B/op        241 allocs/op
BenchmarkNoCache-10        10000             10271 ns/op            6016 B/op        241 allocs/op
BenchmarkNoCache-10        10000             10171 ns/op            6016 B/op        241 allocs/op
BenchmarkNoCache-10        10000             16025 ns/op            6016 B/op        241 allocs/op
BenchmarkNoCache-10        10000              9939 ns/op            6016 B/op        241 allocs/op
BenchmarkNoCache-10        10000             10006 ns/op            6016 B/op        241 allocs/op
BenchmarkNoCache-10        10000              9953 ns/op            6016 B/op        241 allocs/op
BenchmarkNoCache-10        10000             10075 ns/op            6016 B/op        241 allocs/op
BenchmarkNoCache-10        10000              9953 ns/op            6016 B/op        241 allocs/op
BenchmarkNoCache-10        10000             10027 ns/op            6016 B/op        241 allocs/op
PASS
ok      github.com/cloudquery/cloudquery/plugins/destination/postgresql/client  62.879s
```

</details>

#### With caching
Performed by the following command in `plugins/destination/postgresql/client` dir:
```sh
go test \
-test.run=BenchmarkCache \
-test.bench=BenchmarkCache \
-test.count 10 -test.benchmem -test.benchtime 10000x
```

<details><summary>`BenchmarkCache` results</summary>

```
goos: darwin
goarch: arm64
pkg: github.com/cloudquery/cloudquery/plugins/destination/postgresql/client
BenchmarkCache-10          10000               306.9 ns/op             0 B/op          0 allocs/op
BenchmarkCache-10          10000               298.9 ns/op             0 B/op          0 allocs/op
BenchmarkCache-10          10000               518.6 ns/op             0 B/op          0 allocs/op
BenchmarkCache-10          10000               308.7 ns/op             0 B/op          0 allocs/op
BenchmarkCache-10          10000               304.1 ns/op             0 B/op          0 allocs/op
BenchmarkCache-10          10000               331.6 ns/op             0 B/op          0 allocs/op
BenchmarkCache-10          10000               324.6 ns/op             0 B/op          0 allocs/op
BenchmarkCache-10          10000               317.9 ns/op             0 B/op          0 allocs/op
BenchmarkCache-10          10000               327.0 ns/op             0 B/op          0 allocs/op
BenchmarkCache-10          10000               320.7 ns/op             0 B/op          0 allocs/op
PASS
ok      github.com/cloudquery/cloudquery/plugins/destination/postgresql/client  61.732s
```

</details>

Reply:
I added build tags to emulate the diff better in ba2a986d091bdf7346835264c64ae02fcaae1999, the results:

<details><summary>No cache</summary>

```
go test \
-test.run=BenchmarkSQL \
-test.bench=BenchmarkSQL \
-test.count 10 -test.benchmem -test.benchtime 10000x \
-test.timeout 15m
goos: darwin
goarch: arm64
pkg: github.com/cloudquery/cloudquery/plugins/destination/postgresql/client
BenchmarkSQL-10            10000             14682 ns/op            6016 B/op        241 allocs/op
BenchmarkSQL-10            10000             10610 ns/op            6016 B/op        241 allocs/op
BenchmarkSQL-10            10000             10628 ns/op            6016 B/op        241 allocs/op
BenchmarkSQL-10            10000             10675 ns/op            6016 B/op        241 allocs/op
BenchmarkSQL-10            10000             10716 ns/op            6016 B/op        241 allocs/op
BenchmarkSQL-10            10000             10539 ns/op            6016 B/op        241 allocs/op
BenchmarkSQL-10            10000             10784 ns/op            6016 B/op        241 allocs/op
BenchmarkSQL-10            10000             11285 ns/op            6016 B/op        241 allocs/op
BenchmarkSQL-10            10000             10721 ns/op            6016 B/op        241 allocs/op
BenchmarkSQL-10            10000             10842 ns/op            6016 B/op        241 allocs/op
PASS
ok      github.com/cloudquery/cloudquery/plugins/destination/postgresql/client  67.698s
```
</details>

<details><summary>Cache</summary>

```
go test \
-test.run=BenchmarkSQL \
-test.bench=BenchmarkSQL \
-test.count 10 -test.benchmem -test.benchtime 10000x \
-test.timeout 15m \
-tags cache
goos: darwin
goarch: arm64
pkg: github.com/cloudquery/cloudquery/plugins/destination/postgresql/client
BenchmarkSQL-10            10000               339.3 ns/op             0 B/op          0 allocs/op
BenchmarkSQL-10            10000               349.1 ns/op             0 B/op          0 allocs/op
BenchmarkSQL-10            10000               357.8 ns/op             0 B/op          0 allocs/op
BenchmarkSQL-10            10000               351.4 ns/op             0 B/op          0 allocs/op
BenchmarkSQL-10            10000               339.7 ns/op             0 B/op          0 allocs/op
BenchmarkSQL-10            10000               351.0 ns/op             0 B/op          0 allocs/op
BenchmarkSQL-10            10000               345.8 ns/op             0 B/op          0 allocs/op
BenchmarkSQL-10            10000               335.2 ns/op             0 B/op          0 allocs/op
BenchmarkSQL-10            10000               351.1 ns/op             0 B/op          0 allocs/op
BenchmarkSQL-10            10000               341.2 ns/op             0 B/op          0 allocs/op
PASS
ok      github.com/cloudquery/cloudquery/plugins/destination/postgresql/client  66.661s

```
</details>